src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,230 @@
+import { z } from "zod";
+import fetch from "node-fetch";
+import { Message, CommandArg, word } from "../types";
+
+const USAGE: string =
+  "Usage: `?rankup [<username> --target=(username|rank) --language=(py|js|...) --mode={least|each|spread} --limit=(1-8[kyu])]`";
+
+/*
+`?rankup` calculates the number of katas of each rank required for a user to rank up.
+This command is just for fun, however should increase competition between users,
+and (hopefully) increase motivation.
+
+If no username is passed, server nickname of user is taken instead.
+
+Options:
+  <username>   The username to check. Defaults to users discord server nickname
+  --target=    The rank to reach, or a username of a user to overtake. Defaults to users next rank
+  --language=  The language to inspect. Defaults to "overall"
+  --mode=      Calculation mode.
+                  Least favours 1kyus for least overall katas.
+                  Each finds number of katas per rank individually
+                  Spread (default) prioritises high katas, but also distributes points to lower ones
+  --limit=     The max rank to use in calculations ([1-8]kyu)
+*/
+
+enum Mode {
+  Least = "least",
+  Each = "each",
+  Spread = "spread", // default
+}
+const DEFAULTMODE: Mode = Mode.Spread;
+const SPREADWEIGHT: number = 0.6; // used for Spread mode
+
+const LANGS = new Set<string>([
+  "agda",
+  "bf",
+  "c",
+  "cfml",
+  "clojure",
+  "cobol",
+  "coffeescript",
+  "commonlisp",
+  "coq",
+  "cpp",
+  "crystal",
+  "csharp",
+  "d",
+  "dart",
+  "elixir",
+  "elm",
+  "erlang",
+  "factor",
+  "forth",
+  "fortran",
+  "fsharp",
+  "go",
+  "groovy",
+  "haskell",
+  "haxe",
+  "idris",
+  "java",
+  "javascript",
+  "julia",
+  "kotlin",
+  "lean",
+  "lua",
+  "nasm",
+  "nim",
+  "objc",
+  "ocaml",
+  "pascal",
+  "perl",
+  "php",
+  "powershell",
+  "prolog",
+  "purescript",
+  "python",
+  "r",
+  "racket",
+  "raku",
+  "reason",
+  "ruby",
+  "rust",
+  "scala",
+  "shell",
+  "solidarity",
+  "sql",
+  "swift",
+  "typescript",
+  "vb",
+]);
+
+// https://docs.codewars.com/gamification/ranks
+// 3-8 dan added speculatively, based on trend
+const RANKTHRESHOLDS: { [rank: string]: number } = {
+  "7kyu": 20,
+  "6kyu": 76,
+  "5kyu": 229,
+  "4kyu": 643,
+  "3kyu": 1768,
+  "2kyu": 4829,
+  "1kyu": 13147,
+  "1dan": 35759,
+  "2dan": 97225,
+  "3dan": 264324,
+  "4dan": 718593,
+  "5dan": 1953551,
+  "6dan": 5310860,
+  "7dan": 14437910,
+  "8dan": 39250354,
+};
+
+const RANKPOINTS: { [rank: string]: number } = {
+  "8kyu": 2,
+  "7kyu": 3,
+  "6kyu": 8,
+  "5kyu": 21,
+  "4kyu": 55,
+  "3kyu": 149,
+  "2kyu": 404,
+  "1kyu": 1097,
+};
+
+const capitalise = (s: string): string =>
+  s.length ? s[0].toUpperCase() + s.slice(1).toLowerCase() : "";
+
+// Retrieve a users language score, or overall score if language is undefined
+const getUserLangScore = async (user: string, lang?: string): Promise<number | undefined> => {
```

Comment:
I'd define types for the API response and validate the structure, but I'll leave that up to you. It might be too much when you're just learning TypeScript. We can always add the types later.

If you want to try, I'd do something like the following:
```typescript
const getUserLangScore = async (user: string, lang: string = "overall"): Promise<number | undefined> => {
  const response = await fetch("https://www.codewars.com/api/v1/users/" + user);
  if (response.status === 404) return;
  const result = await response.json();
  if (!isRanksResponse(result)) return;
  if (lang !== "overall") {
    // TODO validate `lang`
    return result.ranks.languages[lang].score;
  }
  return result.ranks.overall.score;
};

type RanksResponse = {
  success: true;
  ranks: {
    overall: RankInfo;
    languages: { [lang: string]: RankInfo };
  };
};
type RankInfo = {
  rank: number;
  score: number;
};
const isRanksResponse = (json: unknown): json is RanksResponse => {
  //
};
```

Reply:
Yeah I'll give that a go, I should be able to manage, I think. But I might refactor out the API request to a new file that exports a single function `getCWUserInfo` (or something) and also handles all its own relevant types, if that sounds ok?

Reply:
Yeah, sounds good to me. I'd name it `getUser` in a `codewars` module.

I forgot that [`zod`](https://github.com/colinhacks/zod) is already installed. You can use that to create types with validation.